File: array/longest_non_repeat/longest_non_repeat.py
from typing import Dict


def longest_non_repeat(s: str) -> int:
    start, maxlen = 0, 0
    used_char = {}  # type: Dict[str, int]
    for i, char in enumerate(s):
        if char in used_char and start <= used_char[char]:
            start = used_char[char] + 1
        else:
            maxlen = max(maxlen, i - start + 1)
        used_char[char] = i
    return maxlen


a = "abcabcdefbb"
print("input: ", a)
print("result: ", longest_non_repeat(a))
print("output should be 6")

File: array/longest_non_repeat/longest_non_repeat.go
package main

import "fmt"

func longestNonRepeat(s string) int {
	start, maxlen := 0, 0
	usedChar := make(map[string]int)
	for i, r := range s {
		// r holds a "rune" type, need to convert to string
		char := string(r)
		if val, ok := usedChar[char]; ok && start <= val {
			start = val + 1
			// go has no ternary or built int max
		} else if maxlen <= i+1-start {
			maxlen = i + 1 - start
		}
		usedChar[char] = i
	}
	return maxlen
}

func main() {
	a := "abcabcdefbb"
	fmt.Println("input: ", a)
	fmt.Println("result: ", longestNonRepeat(a))
	fmt.Println("output should be 6")
}
